substrate/runtime/contract/src/gas.rs

```diff
@@ -0,0 +1,124 @@
+// Copyright 2018 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate. If not, see <http://www.gnu.org/licenses/>.
+
+use {Trait, Module};
+use runtime_primitives::traits::{As, CheckedMul, CheckedSub, Zero};
+use staking;
+
+#[must_use]
+#[derive(Debug, PartialEq, Eq)]
+pub enum GasMeterResult {
+	Proceed,
+	OutOfGas,
+}
+
+impl GasMeterResult {
+	pub fn is_out_of_gas(&self) -> bool {
+		match *self {
+			GasMeterResult::OutOfGas => true,
+			GasMeterResult::Proceed => false,
+		}
+	}
+}
+
+pub struct GasMeter<T: Trait> {
+	gas_left: T::Gas,
+	gas_price: T::Balance,
+}
+impl<T: Trait> GasMeter<T> {
+	#[cfg(test)]
+	pub fn with_limit(gas_limit: T::Gas, gas_price: T::Balance) -> GasMeter<T> {
+		GasMeter {
+			gas_left: gas_limit,
+			gas_price,
+		}
+	}
+
+	/// Account for used gas.
+	///
+	/// Returns `OutOfGas` if there is not enough gas or addition of the specified
+	/// amount of gas has lead to overflow. On success returns `Proceed`.
+	pub fn charge(&mut self, amount: T::Gas) -> GasMeterResult {
+		match self.gas_left.checked_sub(&amount) {
+			None => GasMeterResult::OutOfGas,
+			Some(val) if val.is_zero() => GasMeterResult::OutOfGas,
```

Comment:
It's not only about spending last `1` gas, it is in general about spending more gas than available. For example, you have 100 gas left, if you try to consume 200 gas it will lead to `OutOfGas`. 

Although this exact code is shamelessly copied from `kovan-wasm` impl (see `parity/ethcore/wasm/src/runtime.rs`) and, I believe, ultimately is borrowed from the ethereum, I think we can revise it. For example, we can use saturating substraction.

However, I would rather leave it for later. I'll update the PR description.